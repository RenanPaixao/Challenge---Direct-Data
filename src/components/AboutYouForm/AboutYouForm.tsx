import { Center, FormControl, FormLabel, Input, SimpleGrid } from '@chakra-ui/react'
import { useFormik } from 'formik'
import { PropsWithChildren } from 'react'

interface AboutYouValues {
  name: string
  lastName: string
  birthDate: string
  cpf: string
  weight: number
  height: number
  email: string
  phone: string
}

const fieldsConfig: {key: keyof AboutYouValues, type: string, label: string}[]  = [
  {
    key: 'name',
    type: 'text',
    label: 'Nome'
  },
  {
    key: 'lastName',
    type: 'text',
    label: 'Sobrenome'
  },
  {
    key: 'birthDate',
    type: 'date',
    label: 'Data de nascimento'
  },
  {
    key: 'cpf',
    type: 'text',
    label: 'CPF'
  },
  {
    key: 'weight',
    type: 'number',
    label: 'Peso'
  },
  {
    key: 'height',
    type: 'number',
    label: 'Altura'
  },
  {
    key: 'email',
    type: 'email',
    label: 'Email'
  },
  {
    key: 'phone',
    type: 'tel',
    label: 'Telefone'
  }
]

export const AboutYouForm = ({ children }: PropsWithChildren) => {
  const formik = useFormik<AboutYouValues>({
    initialValues: {
      name: '',
      lastName: '',
      birthDate: '',
      cpf: '',
      weight: 0,
      height: 0,
      email: '',
      phone: ''
    },
    onSubmit: values => {
      console.log(values)
    }
  })

  return <Center as={'form'}>
    <SimpleGrid columns={2} spacing={8}>
      {fieldsConfig.map(({ key, type, label }) => {
        return <FormControl key={key}>
          <FormLabel htmlFor={key}>{label}</FormLabel>
          <Input
            type={type}
            name={key}
            aria-label={key}
            onChange={formik.handleChange}
            value={formik.values[key]}
            onBlur={formik.handleBlur}
          />
        </FormControl>
      })}
    </SimpleGrid>
    {children}
  </Center>
}
