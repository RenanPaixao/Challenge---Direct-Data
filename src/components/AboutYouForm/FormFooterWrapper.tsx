import { Flex, FlexProps } from '@chakra-ui/react'
import { PropsWithChildren } from 'react'

export const FormFooterWrapper = ({ children, ...rest }: PropsWithChildren<FlexProps>) => {
  return <Flex justifyContent={'space-between'} w={'100%'} {...rest}>
    {children}
  </Flex>
}
