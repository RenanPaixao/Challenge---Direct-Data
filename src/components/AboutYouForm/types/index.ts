export interface FieldConfig<t> {
  name: keyof t,
  type: string,
  label: string
}
