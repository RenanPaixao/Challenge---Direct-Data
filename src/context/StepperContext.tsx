import { UseStepsReturn } from '@chakra-ui/react'
import { createContext, PropsWithChildren } from 'react'
import { IStep } from '../components/TheStepper/TheStepper.tsx'

interface IProps extends UseStepsReturn {
  steps: IStep[]
}
export const StepperContext = createContext<IProps | null>(null)

StepperContext.displayName = 'StepperContext'

export const StepperProvider = ({ children, ...rest }: PropsWithChildren<IProps>) => {
  return <StepperContext.Provider value={{ ...rest }}>
    {children}
  </StepperContext.Provider>
}

