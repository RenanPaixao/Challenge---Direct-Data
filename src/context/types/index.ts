
export interface AboutYouInformation extends IdentityInformation {
  responsible: IdentityInformation | null
  weight: string
  height: string
  email: string
  phone: string
}

export interface Address {
  street: string,
  number: string,
  complement?: string,
  district: string,
  cep: string,
  city: string,
  state: string
}

export interface IdentityInformation {
  name: string ,
  lastName: string ,
  birthDate: string,
  cpf: string
}
