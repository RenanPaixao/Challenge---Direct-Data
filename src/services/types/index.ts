export interface StudentModel extends IdentityInformationModel {
  responsavel: IdentityInformationModel | null,
  peso: number,
  altura: number,
  email: string,
  telefoneContato: string,
  logradouro: string,
  numero: number,
  complemento: string,
  bairro: string,
  cep: string,
  cidade: string,
  uf: string
}

interface IdentityInformationModel {
  nome: string ,
  sobrenome: string ,
  dataNascimento: string,
  cpf: string
}
