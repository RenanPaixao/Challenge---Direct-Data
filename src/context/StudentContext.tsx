import { createContext, PropsWithChildren, useEffect, useState } from 'react'
import { AboutYouInformation, Address } from './types'
import { persistOnSessionStorage, retrieveFromSessionStorage } from '../utils/persistence.ts'

export interface StudentContextProps {
  aboutYouInformation: AboutYouInformation | null
  setAboutYouInformation: (aboutYouInformation: AboutYouInformation) => void
  address: Address | null
  setAddress: (address: Address) => void
}

export const StudentContext = createContext<StudentContextProps>({
  aboutYouInformation: null,
  setAboutYouInformation: () => {},
  address: null,
  setAddress: () => {}
})

StudentContext.displayName = 'StudentContext'

export const StudentProvider = ({ children }: PropsWithChildren) => {
  const aboutYouStorageKey = 'aboutYouInformation'
  const addressStorageKey = 'address'

  const [aboutYouInformation, setAboutYouInformation] = useState<AboutYouInformation | null>(() => {
    return retrieveFromSessionStorage(aboutYouStorageKey)
  })

  const [address, setAddress] = useState<Address | null>(() => {
    return retrieveFromSessionStorage(addressStorageKey)
  })

  useEffect(() => {
    persistOnSessionStorage(aboutYouStorageKey, aboutYouInformation)
    persistOnSessionStorage(addressStorageKey, address)
  }, [aboutYouInformation, address])

  return <StudentContext.Provider value={{
    aboutYouInformation,
    setAboutYouInformation,
    address,
    setAddress
  }}>
    {children}
  </StudentContext.Provider>
}
