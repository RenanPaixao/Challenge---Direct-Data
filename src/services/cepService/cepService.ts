import axios from 'axios'
import { CepInfo } from '../types'
import { Address } from '../../context/types'

type MappedCep = Omit<Address, 'number'>
const mapCepResponse = (cep: CepInfo): MappedCep  => {
  return {
    cep: cep.cep,
    street: cep.logradouro,
    district: cep.bairro,
    city: cep.localidade,
    state: cep.uf,
    complement: cep.complemento
  }
}

class CepServiceImpl {
  private instance = axios.create({
    baseURL: 'https://viacep.com.br/ws/',
    timeout: 3000
  })

  async getCep(cep: string): Promise<MappedCep> {
    const { data } = await this.instance.get(`${cep}/json`)

    if (!data) {
      throw new Error('CEP n√£o encontrado')
    }

    return mapCepResponse(data)
  }
}

export const cepService = new CepServiceImpl()
