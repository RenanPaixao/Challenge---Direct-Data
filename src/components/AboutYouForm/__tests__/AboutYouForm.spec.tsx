import {
  fillForm,
  FormField,
  renderWithStudentContext
} from '../../../../tests/test-utils.tsx'
import { AboutYouForm } from '../AboutYouForm.tsx'
import { AboutYouInformation } from '../../../context/types'
import { cepService } from '../../../services/cepService/cepService.ts'

type FormValues = Record<keyof Omit<AboutYouInformation, 'responsible'>, FormField>
cepService.getCep = vi.fn().mockReturnValue(Promise.resolve({}))

const formValues: FormValues = {
  name: {
    name: 'name',
    value: 'John',
    role: 'textbox'
  },
  lastName: {
    value: 'Doe',
    role: 'textbox',
    name: 'lastName'
  },
  cpf: {
    value: '09182085056',
    role: 'textbox',
    name: 'cpf'
  },
  email: {
    value: 'a@a.com',
    role: 'textbox',
    name: 'email'
  },
  height: {
    value: '1.80',
    role: 'textbox',
    name: 'height'
  },
  weight: {
    value: '80',
    role: 'textbox',
    name: 'weight'
  },
  birthDate: {
    value: '1990-01-01',
    role: 'Date',
    name: 'birthDate'
  },
  phone: {
    value: '12345678910',
    role: 'textbox',
    name: 'phone'
  }
}

describe('AboutYouForm', () => {
  it('should render', async () => {
    const { container } = renderWithStudentContext(<AboutYouForm />)

    expect(container).toMatchSnapshot()
  })

  it('should render and show responsible fields', async () => {
    const { container } = renderWithStudentContext(<AboutYouForm />)
    await fillForm<FormValues>({
      ...formValues,
      birthDate: { ...formValues.birthDate, value: '2015-12-12' }
    })

    expect(container).toMatchSnapshot()
  })
})

