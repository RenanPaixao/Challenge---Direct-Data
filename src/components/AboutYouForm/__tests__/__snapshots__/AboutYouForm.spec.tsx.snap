// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`AboutYouForm > should render 1`] = `
<div>
  <form
    class="css-b93e9j"
  >
    <div
      class="css-1gkent6"
    >
      <div
        class="chakra-form-control css-6h2d2t"
        role="group"
      >
        <label
          class="chakra-form__label css-g6pte"
          for="name"
          id="field-:r0:-label"
        >
          Nome
        </label>
        <input
          aria-describedby="field-:r0:-helptext"
          aria-label="name"
          class="chakra-input css-1cjy4zv"
          id="field-:r0:"
          name="name"
          type="text"
          value=""
        />
        <div
          class="chakra-form__helper-text css-1fjml25"
          id="field-:r0:-helptext"
        />
      </div>
      <div
        class="chakra-form-control css-6h2d2t"
        role="group"
      >
        <label
          class="chakra-form__label css-g6pte"
          for="lastName"
          id="field-:r1:-label"
        >
          Sobrenome
        </label>
        <input
          aria-describedby="field-:r1:-helptext"
          aria-label="lastName"
          class="chakra-input css-1cjy4zv"
          id="field-:r1:"
          name="lastName"
          type="text"
          value=""
        />
        <div
          class="chakra-form__helper-text css-1fjml25"
          id="field-:r1:-helptext"
        />
      </div>
      <div
        class="chakra-form-control css-6h2d2t"
        role="group"
      >
        <label
          class="chakra-form__label css-g6pte"
          for="birthDate"
          id="field-:r2:-label"
        >
          Data de nascimento
        </label>
        <input
          aria-describedby="field-:r2:-helptext"
          aria-label="birthDate"
          class="chakra-input css-1cjy4zv"
          data-testid="birthDate"
          id="field-:r2:"
          name="birthDate"
          type="date"
          value=""
        />
        <div
          class="chakra-form__helper-text css-1fjml25"
          id="field-:r2:-helptext"
        />
      </div>
      <div
        class="chakra-form-control css-6h2d2t"
        role="group"
      >
        <label
          class="chakra-form__label css-g6pte"
          for="cpf"
          id="field-:r3:-label"
        >
          CPF
        </label>
        <input
          aria-describedby="field-:r3:-helptext"
          aria-label="cpf"
          class="chakra-input css-1cjy4zv"
          id="field-:r3:"
          name="cpf"
          type="text"
          value=""
        />
        <div
          class="chakra-form__helper-text css-1fjml25"
          id="field-:r3:-helptext"
        />
      </div>
      <div
        class="chakra-form-control css-6h2d2t"
        role="group"
      >
        <label
          class="chakra-form__label css-g6pte"
          for="weight"
          id="field-:r4:-label"
        >
          Peso
        </label>
        <input
          aria-describedby="field-:r4:-helptext"
          aria-label="weight"
          class="chakra-input css-1cjy4zv"
          id="field-:r4:"
          name="weight"
          type="text"
          value=""
        />
        <div
          class="chakra-form__helper-text css-1fjml25"
          id="field-:r4:-helptext"
        >
          Exemplo: 70,00
        </div>
      </div>
      <div
        class="chakra-form-control css-6h2d2t"
        role="group"
      >
        <label
          class="chakra-form__label css-g6pte"
          for="height"
          id="field-:r5:-label"
        >
          Altura
        </label>
        <input
          aria-describedby="field-:r5:-helptext"
          aria-label="height"
          class="chakra-input css-1cjy4zv"
          id="field-:r5:"
          name="height"
          type="text"
          value=""
        />
        <div
          class="chakra-form__helper-text css-1fjml25"
          id="field-:r5:-helptext"
        >
          Exemplo: 1,70
        </div>
      </div>
      <div
        class="chakra-form-control css-6h2d2t"
        role="group"
      >
        <label
          class="chakra-form__label css-g6pte"
          for="email"
          id="field-:r6:-label"
        >
          Email
        </label>
        <input
          aria-describedby="field-:r6:-helptext"
          aria-label="email"
          class="chakra-input css-1cjy4zv"
          id="field-:r6:"
          name="email"
          type="email"
          value=""
        />
        <div
          class="chakra-form__helper-text css-1fjml25"
          id="field-:r6:-helptext"
        />
      </div>
      <div
        class="chakra-form-control css-6h2d2t"
        role="group"
      >
        <label
          class="chakra-form__label css-g6pte"
          for="phone"
          id="field-:r7:-label"
        >
          Telefone
        </label>
        <input
          aria-describedby="field-:r7:-helptext"
          aria-label="phone"
          class="chakra-input css-1cjy4zv"
          id="field-:r7:"
          name="phone"
          type="tel"
          value=""
        />
        <div
          class="chakra-form__helper-text css-1fjml25"
          id="field-:r7:-helptext"
        />
      </div>
    </div>
    <div
      class="css-um72kk"
    >
      <button
        class="chakra-button css-ez23ye"
        type="submit"
      >
        Avan√ßar
      </button>
    </div>
  </form>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;

exports[`AboutYouForm > should render and show responsible fields 1`] = `
<div>
  <form
    class="css-b93e9j"
  >
    <div
      class="css-1gkent6"
    >
      <div
        class="chakra-form-control css-6h2d2t"
        role="group"
      >
        <label
          class="chakra-form__label css-g6pte"
          for="name"
          id="field-:r8:-label"
        >
          Nome
        </label>
        <input
          aria-describedby="field-:r8:-helptext"
          aria-label="name"
          class="chakra-input css-1cjy4zv"
          id="field-:r8:"
          name="name"
          type="text"
          value="John"
        />
        <div
          class="chakra-form__helper-text css-1fjml25"
          id="field-:r8:-helptext"
        />
      </div>
      <div
        class="chakra-form-control css-6h2d2t"
        role="group"
      >
        <label
          class="chakra-form__label css-g6pte"
          for="lastName"
          id="field-:r9:-label"
        >
          Sobrenome
        </label>
        <input
          aria-describedby="field-:r9:-helptext"
          aria-label="lastName"
          class="chakra-input css-1cjy4zv"
          id="field-:r9:"
          name="lastName"
          type="text"
          value="Doe"
        />
        <div
          class="chakra-form__helper-text css-1fjml25"
          id="field-:r9:-helptext"
        />
      </div>
      <div
        class="chakra-form-control css-6h2d2t"
        role="group"
      >
        <label
          class="chakra-form__label css-g6pte"
          for="birthDate"
          id="field-:ra:-label"
        >
          Data de nascimento
        </label>
        <input
          aria-describedby="field-:ra:-helptext"
          aria-label="birthDate"
          class="chakra-input css-1cjy4zv"
          data-testid="birthDate"
          id="field-:ra:"
          name="birthDate"
          type="date"
          value="2015-12-12"
        />
        <div
          class="chakra-form__helper-text css-1fjml25"
          id="field-:ra:-helptext"
        />
      </div>
      <div
        class="chakra-form-control css-6h2d2t"
        role="group"
      >
        <label
          class="chakra-form__label css-g6pte"
          for="cpf"
          id="field-:rb:-label"
        >
          CPF
        </label>
        <input
          aria-describedby="field-:rb:-helptext"
          aria-label="cpf"
          class="chakra-input css-1cjy4zv"
          id="field-:rb:"
          name="cpf"
          type="text"
          value=""
        />
        <div
          class="chakra-form__helper-text css-1fjml25"
          id="field-:rb:-helptext"
        />
      </div>
      <div
        class="chakra-form-control css-6h2d2t"
        role="group"
      >
        <label
          class="chakra-form__label css-g6pte"
          for="weight"
          id="field-:rc:-label"
        >
          Peso
        </label>
        <input
          aria-describedby="field-:rc:-helptext"
          aria-label="weight"
          class="chakra-input css-1cjy4zv"
          id="field-:rc:"
          name="weight"
          type="text"
          value=""
        />
        <div
          class="chakra-form__helper-text css-1fjml25"
          id="field-:rc:-helptext"
        >
          Exemplo: 70,00
        </div>
      </div>
      <div
        class="chakra-form-control css-6h2d2t"
        role="group"
      >
        <label
          class="chakra-form__label css-g6pte"
          for="height"
          id="field-:rd:-label"
        >
          Altura
        </label>
        <input
          aria-describedby="field-:rd:-helptext"
          aria-label="height"
          class="chakra-input css-1cjy4zv"
          id="field-:rd:"
          name="height"
          type="text"
          value=""
        />
        <div
          class="chakra-form__helper-text css-1fjml25"
          id="field-:rd:-helptext"
        >
          Exemplo: 1,70
        </div>
      </div>
      <div
        class="chakra-form-control css-6h2d2t"
        role="group"
      >
        <label
          class="chakra-form__label css-g6pte"
          for="email"
          id="field-:re:-label"
        >
          Email
        </label>
        <input
          aria-describedby="field-:re:-helptext"
          aria-label="email"
          class="chakra-input css-1cjy4zv"
          id="field-:re:"
          name="email"
          type="email"
          value="a@a.com"
        />
        <div
          class="chakra-form__helper-text css-1fjml25"
          id="field-:re:-helptext"
        />
      </div>
      <div
        class="chakra-form-control css-6h2d2t"
        data-focus=""
        role="group"
      >
        <label
          class="chakra-form__label css-g6pte"
          data-focus=""
          for="phone"
          id="field-:rf:-label"
        >
          Telefone
        </label>
        <input
          aria-describedby="field-:rf:-helptext"
          aria-label="phone"
          class="chakra-input css-1cjy4zv"
          id="field-:rf:"
          name="phone"
          type="tel"
          value=""
        />
        <div
          class="chakra-form__helper-text css-1fjml25"
          id="field-:rf:-helptext"
        />
      </div>
      <div
        class="chakra-form-control css-6h2d2t"
        role="group"
      >
        <label
          class="chakra-form__label css-g6pte"
          for="responsibleName"
          id="field-:rg:-label"
        >
          Nome do Respons√°vel
        </label>
        <input
          aria-label="responsibleName"
          class="chakra-input css-1cjy4zv"
          id="field-:rg:"
          name="responsibleName"
          type="text"
          value=""
        />
      </div>
      <div
        class="chakra-form-control css-6h2d2t"
        role="group"
      >
        <label
          class="chakra-form__label css-g6pte"
          for="responsibleLastName"
          id="field-:rh:-label"
        >
          Sobrenome do Respons√°vel
        </label>
        <input
          aria-label="responsibleLastName"
          class="chakra-input css-1cjy4zv"
          id="field-:rh:"
          name="responsibleLastName"
          type="text"
          value=""
        />
      </div>
      <div
        class="chakra-form-control css-6h2d2t"
        role="group"
      >
        <label
          class="chakra-form__label css-g6pte"
          for="responsibleBirthDate"
          id="field-:ri:-label"
        >
          Data de nascimento do Respons√°vel
        </label>
        <input
          aria-label="responsibleBirthDate"
          class="chakra-input css-1cjy4zv"
          data-testid="responsibleBirthDate"
          id="field-:ri:"
          name="responsibleBirthDate"
          type="date"
          value=""
        />
      </div>
      <div
        class="chakra-form-control css-6h2d2t"
        role="group"
      >
        <label
          class="chakra-form__label css-g6pte"
          for="responsibleCpf"
          id="field-:rj:-label"
        >
          CPF do Respons√°vel
        </label>
        <input
          aria-label="responsibleCpf"
          class="chakra-input css-1cjy4zv"
          id="field-:rj:"
          name="responsibleCpf"
          type="text"
          value=""
        />
      </div>
    </div>
    <div
      class="css-um72kk"
    >
      <button
        class="chakra-button css-ez23ye"
        type="submit"
      >
        Avan√ßar
      </button>
    </div>
  </form>
  <span
    hidden=""
    id="__chakra_env"
  />
</div>
`;
